tools/common/FileCopier.cs

```diff
@@ -0,0 +1,249 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+
+namespace Xamarin.Bundler {
+	public static class FileCopier 
+	{
+		enum CopyFileFlags : uint {
+			ACL = 1 << 0,
+			Stat = 1 << 1,
+			Xattr = 1 << 2,
+			Data = 1 << 3,
+			Security = Stat | ACL,
+			Metadata = Security | Xattr,
+			All = Metadata | Data,
+
+			Recursive = 1 << 15,
+			NoFollow_Src = 1 << 18,
+			NoFollow_Dst = 1 << 19,
+			Unlink = 1 << 21,
+			Nofollow = NoFollow_Src | NoFollow_Dst,
+			Clone = 1 << 24,
+		}
+
+		enum CopyFileState : uint {
+			StatusCB = 6,
+		}
+
+		enum CopyFileStep {
+			Start = 1,
+			Finish = 2,
+			Err = 3,
+			Progress = 4,
+		}
+
+		enum CopyFileResult {
+			Continue = 0,
+			Skip = 1,
+			Quit = 2,
+		}
+
+		enum CopyFileWhat {
+			Error = 0,
+			File = 1,
+			Dir = 2,
+			DirCleanup = 3,
+			CopyData = 4,
+			CopyXattr = 5,
+		}
+
+		[DllImport ("/usr/lib/libSystem.dylib")]
+		static extern IntPtr copyfile_state_alloc ();
+
+		[DllImport ("/usr/lib/libSystem.dylib")]
+		static extern int copyfile_state_free (IntPtr state);
+
+		[DllImport ("/usr/lib/libSystem.dylib")]
+		static extern int copyfile_state_set (IntPtr state, CopyFileState flag, IntPtr value);
+
+		delegate CopyFileResult CopyFileCallbackDelegate (CopyFileWhat what, CopyFileStep stage, IntPtr state, string src, string dst, IntPtr ctx);
+
+		[DllImport ("/usr/lib/libSystem.dylib", SetLastError = true)]
+		static extern int copyfile (string @from, string @to, IntPtr state, CopyFileFlags flags);
+
+		// This code is shared between our packaging tools (mmp\mtouch) and msbuild tasks
+#if MMP || MTOUCH
+		public static void Log (int min_verbosity, string format, params object[] args) => Driver.Log (min_verbosity, format, args);
+		public static bool Force => Driver.Force;
+		public static Exception CreateError (int code, string message, params object[] args) => ErrorHelper.CreateError (code, message, args);
+#else
+		public static void Log (int min_verbosity, string format, params object[] args) => Console.WriteLine (format, args);
+		public static bool Force => false;
+		public static Exception CreateError (int code, string message, params object[] args) => throw new Exception ($"{code} {string.Format (message, args)}");
```

Comment:
I think the logging/error code should use MSBuild's infrastructure.

Reply:
https://github.com/xamarin/xamarin-macios/issues/5232